lib/ex_uid2/dsp.ex

```diff
@@ -0,0 +1,78 @@
+defmodule ExUid2.Dsp do
+  alias ExUid2.Encryption
+  alias ExUid2.Keyring
+  alias ExUid2.Api
+  use GenServer
+
+  @table_name __MODULE__
+  @keyring :uid2_decryption_keyring
+  @refresh_rate_ms 5000
+
+  require Logger
+
+  def start_link(opts \\ []) do
+    :ets.new(@table_name, [:named_table, {:read_concurrency, true}, :public])
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  def child_spec(_opts) do
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, []}
+    }
+  end
+
+  @impl GenServer
+  def init(_opts) do
+    send(self(), :refresh)
+    {:ok, %{fetch_keyring_attempts: 0}}
+  end
+
+  @impl GenServer
+  def handle_info(:refresh, state) do
+    case Api.Sharing.fetch_keyring() do
+      {:ok, fresh_keyring} ->
+        :ets.insert(@table_name, {@keyring, fresh_keyring})
+        Process.send_after(self(), :refresh, @refresh_rate_ms)
+        {:noreply, put_in(state.fetch_keyring_attempts, 0)}
+
+      error ->
+        nb_attempts = state.fetch_keyring_attempts
+        retry_time_ms = round(:math.pow(2, nb_attempts))
+
+        Logger.warning(
+          "Failed to fetch keyring: #{inspect(error)}. Retrying in #{retry_time_ms} ms."
+        )
+
+        Process.send_after(self(), :refresh, retry_time_ms)
+        # Exponential backoff maxes out at a ~32 seconds interval
+        nb_attempts = min(nb_attempts + 1, 15)
```

Comment:
For the sake of clarity, I would store the true number of attempts in the state, and instead call `min/2` in the call to `:math.pow(2, ...)`

Reply:
I also think that the first 8ish retries will be a bit aggressively fast, though I don't really know what to suggest as an alternative.

Reply:
I changed it to be a base of 500ms and doubling up to 32s.